File: BOJ_백준/12789 도키도키 간식드리미 (스택, 큐, 덱).cpp
#include <iostream>
#include <stack>
#include <queue>

using namespace std;

int main() {
    int N, num;
    stack <int> st;
    queue <int> que;
    cin >> N;
    while (N--) {
        cin >> num;
        que.push(num);
    }

    int i = 1;
    while (!que.empty()) {
        if (que.front() == i) {
            i++;
            que.pop();
        }
        else {
            st.push(que.front());
            que.pop();
        }
        while (!st.empty() && st.top() == i) {
            st.pop();
            i++;
        }
    }

    if (st.empty())
        cout << "Nice";
    else
        cout << "Sad";
    
    return 0;
}

File: BOJ_백준/12789 도키도키 간식드리미 (스택, 큐, 덱).py
import sys
import collections

input = sys.stdin.readline
N = int(input())
que = collections.deque(map(int, input().split()))
st = collections.deque()

i = 1
while que:
    if que[0] == i:
        que.popleft()
        i += 1
    else:
        st.append(que.popleft())
    while st and st[-1] == i:
        st.pop()
        i += 1
        
print("Nice" if not st else "Sad")
